### ISSUE TITLE
[Bug]: Unable to update the batch_normalization layer moving_mean/moving_variance of keras 

### ISSUE THREAD

**Issue by zizhaozhang** [tensorflow/tensorflow#16102]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: Tensorflow 1.4.1
- **Bazel version (if compiling from source)**: N/A
- **Python version**: Python 3.5
- **CUDA/cuDNN version**: 8.0
- **GPU model and memory**: 1080 Ti
- **Exact command to reproduce**: see below

It is following up this [issue](https://github.com/tensorflow/tensorflow/issues/15367#issuecomment-357088739) about training models by mixing Tensorflow with Keras using this style, exampled [here](https://blog.keras.io/keras-as-a-simplified-interface-to-tensorflow-tutorial.html). Basically, using Inception architectures in keras.applications.InceptionV3 and training using sess.run() with a placeholder K.learning_phase() to indicate the training/testing mode.

_At beginning, I found the evaluation accuracy in testing mode (loaded back from a checkpoint) is different from the results when evaluation on the fly during training. But when I evaluated using training mode, it can give me reasonable but not good results. So I knew the problem happened in batch normalization._

**I tried either using tf.keras and keras independent version (2.1.2). But both not work.**

After some investigations, I found there should an issue of Keras of using batch normalization. I track the moving_mean and moving_variance and found they never update (maintaining the initial values, zero and one). In tf. keras, I notice its batchnorm inherits tf.layers.BatchNormalization (see [here](https://github.com/tensorflow/tensorflow/blob/ac8e67399d75edce6a9f94afaa2adb577035966e/tensorflow/python/keras/_impl/keras/layers/normalization.py#L26)). However, by checking the tutorial of this tf.layers.BatchNormalization,
  _It required to add a **update_ops** in optimizer. But it never mentioned in Keras (which should be 
  clarified)._ 
I did not see Keras using update_ops anywhere. **So i believe if you want to fine-tune a keras predefined model in applications, you never be able to update your moving_mean and moving_variance for your new data**.

**Note that I also tried keras independent version. It never worked. moving_mean and moving_variance are always not changed. I  tracked the value of batch_mean which is used to update moving_mean. batch_mean has values but not on moving_mean.**

In addition, I found moving_mean and moving_variance are in tf.trainable_variables() when using keras but not in tf.keras. I am not sure if this matters.

Here is an example code

```
image = keras.preprocessing.image
def preprocess_input(x):
    # the same as keras.applications.inception_v3
    with tf.name_scope('preprocess_input'):
        x /= 255.
        x -= 0.5
        x *= 2.
        return x

def get_main_network(name, input_tensor, use_weights=False):

    processed_tensor = preprocess_input(input_tensor)

    if name == 'inception':
        base_model = keras.applications.InceptionV3(include_top=True,
                                                    weights='imagenet' if use_weights else None,
                                                    pooling='avg',
                                                    input_tensor=processed_tensor)
    
    model = keras.models.Model(inputs=base_model.input, outputs=base_model.output)
 
    return model

img_shape=[299,299]
img = tf.placeholder(tf.float32, shape=[None]+img_shape+[3])

with tf.name_scope('model'):
    model = get_main_network('inception', input_tensor=img, use_weights=False)
    output = model.output
    logit = tf.cast(tf.argmax(output, axis=1), np.float32)

# define loss
with tf.name_scope('cross_entropy'):
    cross_entropy_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=label, logits=output)  

# define optimizer
with tf.name_scope('learning_rate'):
    global_step = tf.Variable(0, name='global_step', trainable=False)
    learning_rate = tf.train.exponential_decay(opt.learning_rate, global_step,
                                        iter_epoch*opt.lr_decay_epoch, opt.lr_decay, staircase=True)
    optimizer = tf.train.MomentumOptimizer(learning_rate=learning_rate, momentum=0.9) #.minimize(cross_entropy_loss, global_step=global_step)

''' **This matters a lot but not working for independent keras version'**''
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
with tf.control_dependencies(update_ops):
    train_step = optimizer.minimize(cross_entropy_loss, global_step=global_step)

''' Initialization '''
init_op = tf.variables_initializer([]) # a fake one, variables already initialized in keras
sess.run(init_op)

img_path = 'elephant.jpg'
imgs= image.load_img(img_path, target_size=img_shape)
x = image.img_to_array(imgs)
x = np.expand_dims(x, axis=0)
saver = tf.train.Saver(max_to_keep=20) # must be added in the end
with sess.as_default():
    
    feed_dict = {   
                        img: x_batch,
                        label: y_batch,
                        K.learning_phase(): True
                    }
        _, loss = sess.run([train_step, 
                                    cross_entropy_loss, 
                                    ], feed_dict=feed_dict)

```
@tensorflowbutler responsed

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version
Exact command to reproduce

**Comment by ymcasky** [comment]
Dear all, 
I have other example code shows keras's BatchNormalization doesn't work as following.

```

import tensorflow as tf
import numpy as np


#from tensorflow.python.keras import  backend
from tensorflow import keras
K = keras.backend

from tensorflow.python.keras.layers import  Input

from tensorflow.python.keras.layers import MaxPooling2D
from tensorflow.python.keras.layers import BatchNormalization
from tensorflow.python.keras.layers import Conv2D, Dense, Flatten
from tensorflow.python.keras.layers import Activation
from tensorflow.python.keras.models import Model

print(tf.__version__)
# Load Data
import cifar10

img_size = 32
num_channels = 3
num_classes = 10
cifar10.maybe_download_and_extract()


class_names = cifar10.load_class_names()


images_train, cls_train, labels_train = cifar10.load_training_data()

images_test, cls_test, labels_test = cifar10.load_test_data()

print("Size of:")
print("- Training-set:\t\t{}".format(len(images_train)))
print("- Test-set:\t\t{}".format(len(images_test)))



img_size_cropped = 24
#%%
def pre_process_image(image, training):
    # This function takes a single image as input,
    # and a boolean whether to build the training or testing graph.
    
    if training:
        # For training, add the following to the TensorFlow graph.

        # Randomly crop the input image.
        image = tf.random_crop(image, size=[img_size_cropped, img_size_cropped, num_channels])

        # Randomly flip the image horizontally.
        image = tf.image.random_flip_left_right(image)
        
        # Randomly adjust hue, contrast and saturation.
        image = tf.image.random_hue(image, max_delta=0.05)
        image = tf.image.random_contrast(image, lower=0.3, upper=1.0)
        image = tf.image.random_brightness(image, max_delta=0.2)
        image = tf.image.random_saturation(image, lower=0.0, upper=2.0)

        # Some of these functions may overflow and result in pixel
        # values beyond the [0, 1] range. It is unclear from the
        # documentation of TensorFlow 0.10.0rc0 whether this is
        # intended. A simple solution is to limit the range.

        # Limit the image pixels between [0, 1] in case of overflow.
        image = tf.minimum(image, 1.0)
        image = tf.maximum(image, 0.0)
    else:
        # For training, add the following to the TensorFlow graph.

        # Crop the input image around the centre so it is the same
        # size as images that are randomly cropped during training.
        image = tf.image.resize_image_with_crop_or_pad(image,
                                                       target_height=img_size_cropped,
                                                       target_width=img_size_cropped)

    return image

def pre_process(images, training):
    # Use TensorFlow to loop over all the input images and call
    # the function above which takes a single image as input.
    images = tf.map_fn(lambda image: pre_process_image(image, training), images)

    return images

#%%
def _bn_relu(input_layer):
    """Helper to build a BN -> relu block
    """
    norm = BatchNormalization(axis=3)(input_layer)
    return Activation("relu")(norm)
#%%

x = tf.placeholder(tf.float32, shape=[None, img_size, img_size, num_channels], name='x')

distorted_images = pre_process(images=x, training=True)

y_true = tf.placeholder(tf.float32, shape=[None, num_classes], name='y_true')

y_true_cls = tf.argmax(y_true, axis=1)

def create_network(training):
    # Wrap the neural network in the scope named 'network'.
    # Create new variables during training, and re-use during testing.
    with tf.variable_scope('network', reuse=not training):

        images = x        
        images = pre_process(images=images, training=training)
        
        inputs = Input(tensor=images)

        net_ks = Conv2D(kernel_size=5, strides=1, filters=64, padding='same',
             activation='linear', name='layer_conv1')(inputs)

        net_ks = _bn_relu(net_ks)
        net_ks = MaxPooling2D(pool_size=2, strides=2)(net_ks)


        net_ks = Conv2D(kernel_size=5, strides=1, filters=64, padding='same',
        activation='relu', name='layer_conv2')(net_ks) 
        net_ks = MaxPooling2D(pool_size=2, strides=2)(net_ks)

        
        net_ks = Flatten()(net_ks)
        
        net_ks = Dense(256, activation='relu', name='layer_fc1')(net_ks)
        net_ks = Dense(128, activation='relu', name='layer_fc2')(net_ks)
        preds_ks = Dense(num_classes, activation='linear')(net_ks)
        
        preds_softmax = tf.nn.softmax(preds_ks)
        step1 = tf.cast(y_true, tf.float32) * tf.log(preds_softmax)
        step2 = -tf.reduce_sum(step1, reduction_indices=[1])
        loss = tf.reduce_mean(step2)       # loss
        
        model = Model(inputs=inputs, outputs=preds_ks)

    return preds_softmax, loss, model

#%% Create Neural Network for Training Phase
global_step = tf.Variable(initial_value=0,
                          name='global_step', trainable=False)

y_pred, loss, model = create_network(training=True)

model.summary()

optimizer = tf.train.AdamOptimizer(learning_rate=1e-4).minimize(loss, global_step=global_step)
#%% Create Neural Network for Test Phase / Inference
#y_pred, _ , _= create_network(training=False)
y_pred_cls = tf.argmax(y_pred, axis=1)
correct_prediction = tf.equal(y_pred_cls, y_true_cls)
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

session = tf.Session()
session.run(tf.global_variables_initializer())
train_batch_size = 64

def random_batch():
    # Number of images in the training-set.
    num_images = len(images_train)

    # Create a random index.
    idx = np.random.choice(num_images,
                           size=train_batch_size,
                           replace=False)

    # Use the random index to select random images and labels.
    x_batch = images_train[idx, :, :, :]
    y_batch = labels_train[idx, :]

    return x_batch, y_batch

def optimize(num_iterations):

    for i in range(num_iterations):

        x_batch, y_true_batch = random_batch()


        feed_dict_train = {x: x_batch,
                           y_true: y_true_batch, K.learning_phase(): 1}


        i_global, _ = session.run([global_step, optimizer],
                                  feed_dict=feed_dict_train)

        # Print status to screen every 100 iterations (and last).
        if (i_global % 100 == 0) or (i == num_iterations - 1):
            # Calculate the accuracy on the training-batch.
            feed_dict_train = {x: x_batch,
                           y_true: y_true_batch, K.learning_phase(): 0}
            batch_acc = session.run(accuracy,
                                    feed_dict=feed_dict_train)

            # Print status.
            msg = "Global Step: {0:>6}, Training Batch Accuracy(phase 0): {1:>6.1%}"
            print(msg.format(i_global, batch_acc))
            
            feed_dict_train = {x: x_batch,
                           y_true: y_true_batch, K.learning_phase(): 1}
            batch_acc = session.run(accuracy,
                                    feed_dict=feed_dict_train)

            # Print status.
            msg = "Global Step: {0:>6}, Training Batch Accuracy(phase 1): {1:>6.1%}"
            print(msg.format(i_global, batch_acc))  


#%%
optimize(num_iterations=10000)
```
Please check this bug. Thanks a lot!!

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by bhack** [comment]
Check also https://github.com/tensorflow/tensorflow/issues/16455

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by fchollet** [comment]
> I did not see Keras using update_ops anywhere

That is by design. Global collections (and global states in general) are prone to various issues (like most global variables in software engineering) and should be avoided.

You can retrieve the updates created by your BatchNormalization layer via:

```python
updates = layer.updates
```

Additionally you can retrieve the aggregated updates of all stateful layers in a Keras model via:

```python
updates = model.updates
```

If you are writing your own custom training loops, you will have to run these updates as part of the call to `sess.run()`. But do note that tf.keras provides a built-in training loop in the form of `model.fit()`, as well as primitive methods for building your own custom training loops (in particular `model.train_on_batch()`  and `model.predict_on_batch()`). If you are using Keras you should generally never have to manually manage your model's updates.

**Comment by tanzhenyu** [comment]
Hi @zizhaozhang and @ymcasky . In both your examples, can you try adding K.set_learning_phase(1) at the entry point of create_main_network() and see if it works?

**Comment by chrishkchris** [comment]
Encountered exactly the same problem. It is not convenience to have the moving average and variance in the batch normalization layer updated, when we do a customized training loop by utilizing both tensorflow's graph and keras model. When we do customized training loop, using model.train_on_batch() and model.predict_on_batch() may not be enough, especially for those working in distributed training. Have tried the following three method to indicate the different between training and evaluation but still never update the moving average.
1. tf.keras.backend.set_learning_phase
2. tf.keras.backend.uses_learning_phase 
3. model(input,training=training)

**Comment by chrishkchris** [comment]
Is the problem caused by not setting the tensorflow session of the keras backend? 
So, can the problem of moving_mean update be solved by adding a line
tf.keras.backend.set_session(session)
after the line
session = tf.Session() ?
